actors/builtin/multisig/multisig_actor.go

```diff
@@ -357,42 +366,59 @@ func (a Actor) ChangeNumApprovalsThreshold(rt vmr.Runtime, params *ChangeNumAppr
 
 func (a Actor) approveTransaction(rt vmr.Runtime, txnID TxnID, proposalHash []byte, checkHash bool) (bool, []byte, exitcode.ExitCode) {
 	var st State
-	var txn Transaction
-	rt.State().Transaction(&st, func() interface{} {
-		var err error
-		txn, err = st.getPendingTransaction(adt.AsStore(rt), txnID)
-		if err != nil {
-			rt.Abortf(exitcode.ErrNotFound, "failed to get transaction for approval: %v", err)
-		}
-		// abort duplicate approval
-		for _, previousApprover := range txn.Approved {
-			if previousApprover == rt.Message().Caller() {
-				rt.Abortf(exitcode.ErrIllegalState, "already approved this message")
-			}
-		}
+	txn := a.getTransaction(rt, txnID, proposalHash, checkHash)
 
-		// confirm the hashes match
-		if checkHash {
-			calculatedHash, err := ComputeProposalHash(&txn, rt.Syscalls().HashBlake2b)
-			if err != nil {
-				rt.Abortf(exitcode.ErrIllegalState, "failed to compute proposal hash: %v", err)
-			}
-			if proposalHash != nil && !bytes.Equal(proposalHash, calculatedHash[:]) {
-				rt.Abortf(exitcode.ErrIllegalState, "hash does not match proposal params")
-			}
+	// abort duplicate approval
+	for _, previousApprover := range txn.Approved {
+		if previousApprover == rt.Message().Caller() {
+			rt.Abortf(exitcode.ErrIllegalState, "already approved this message")
 		}
+	}
 
+	// add the caller to the list of approvers
+	rt.State().Transaction(&st, func() interface{} {
 		// update approved on the transaction
 		txn.Approved = append(txn.Approved, rt.Message().Caller())
-		if err = st.putPendingTransaction(adt.AsStore(rt), txnID, txn); err != nil {
+		if err := st.putPendingTransaction(adt.AsStore(rt), txnID, txn); err != nil {
 			rt.Abortf(exitcode.ErrIllegalState, "failed to put transaction for approval: %v", err)
 		}
 		return nil
 	})
 
+	return executeTransactionIfApproved(rt,st, txnID, txn)
+}
+
+func (a Actor) getTransaction(rt vmr.Runtime, txnID TxnID, proposalHash []byte, checkHash bool) Transaction {
```

Comment:
This factoring results in an additional state loading, of state that's already loaded before this. We want to minimise the number of state transactions or reads.

In `Approve `, move the call to this method back inside the state transaction that method already has, and pass the state here rather than re-load it. Similarly in `approveTransaction`. In fact, `Approve` ends up loading the same transaction twice because it calls `approveTransaction`.

Try to refactor so all the flows have the shape:
- in first state transaction, load all necessary data, call isSigner (which may write a newly-resolved address to state), mutate approvers
- outside of a state transaction check whether txn is ready to execute and execute if so
- iff it was executed, in a second state transaction, remove it from state

Reply:
@anorth What is the cost of a State transaction ? Does it have to load the entire state from disk ?

Reply:
Also, I've finished this refactoring.

Reply:
We don't quite control what it does. Maybe the encoded block is cached. Maybe the decoded block is even cached too, but probably not. The cost is primarily en/decoding, and maybe disk access.